app/gui2/src/components/CodeEditor.vue

```diff
@@ -0,0 +1,66 @@
+<script setup lang="ts">
+import { ref, watchEffect } from 'vue'
+import { useWindowEvent } from '@/util/events'
+
+const emit = defineEmits<{
+  codeUpdate: [code: string]
+}>()
+
+const content = ref('main = \n    2 + 2')
+const shown = ref(false)
+
+watchEffect(() => {
+  emit('codeUpdate', content.value)
+})
+
+const textArea = ref<InstanceType<typeof HTMLTextAreaElement> | null>(null)
+
+useWindowEvent('keydown', (e) => {
+  const graphEditorInFocus = document.activeElement === document.body
+  const codeEditorInFocus =
+    document.activeElement === document.querySelector('.CodeEditor > textarea')
```

Comment:
That `querySelector` is not necessary here. You already have access to that DOM node through the `textArea` ref.

```suggestion
    document.activeElement === textArea.value
```

For now that's enough, because we have a textarea which is always a leaf DOM node. But in more general cases (e.g. component browser, or more complex editor), you would rather check if the focus is anywhere within that element. Say `rootElement` is a `ref` containing the `.CodeEditor` element, it would look like so:

```ts
    rootElement.value?.contains(document.activeElement)
```